# Визуализация данных с использованием Plotly

Plotly – это библиотека с открытым исходным кодом для Python и R, которая отлично подходит для создания красивых и интерактивных визуализаций.

Plotly — библиотека для визуализации данных, состоящая из нескольких частей:
- Front-End на JS
- Back-End на Python (за основу взята библиотека Seaborn)
- Back-End на R

Plotly позволяет строить интерактивные визуализации. 

Документация:

- [Plotly](https://plotly.com/)

- [Plotly.Express](https://plotly.com/python/plotly-express/)

- [Plotly.Dash](https://dash.plotly.com/) - построение дашбордов



In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import numpy as np

In [2]:
df = pd.read_csv('../task_1_Docker/datasets/aggrigation_logs_per_week.csv')
df

,courseid,userid,num_week,s_all,s_all_avg,s_course_viewed,s_course_viewed_avg,s_q_attempt_viewed,s_q_attempt_viewed_avg,s_a_course_module_viewed,...,s_a_submission_status_viewed_avg,NameR_Level,Name_vAtt,Depart,Name_OsnO,Name_FormOPril,LevelEd,Num_Sem,Kurs,Date_vAtt
0,71262,34527,6,9,9,4,4,0,0,0,...,0,3,Экзамен,22,1,1,1,2,2,18.06.2022
1,71262,34527,7,0,"4,5",0,2,0,0,0,...,0,3,Экзамен,22,1,1,1,2,2,18.06.2022
2,71262,34527,8,0,3,0,"1,3333",0,0,0,...,0,3,Экзамен,22,1,1,1,2,2,18.06.2022
3,71262,34527,9,0,"2,25",0,1,0,0,0,...,0,3,Экзамен,22,1,1,1,2,2,18.06.2022
4,71262,34527,10,0,"1,8",0,"0,8",0,0,0,...,0,3,Экзамен,22,1,1,1,2,2,18.06.2022
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
414523,89948,28911,25,0,"2,15",0,"0,25",0,"1,3",0,...,"0,05",5,Экзамен,9,1,1,1,4,3,17.06.2022
414524,89948,28911,26,0,"2,0476",0,"0,2381",0,"1,2381",0,...,"0,0476",5,Экзамен,9,1,1,1,4,3,17.06.2022
414525,89948,28911,27,0,"1,9545",0,"0,2273",0,"1,1818",0,...,"0,0455",5,Экзамен,9,1,1,1,4,3,17.06.2022
414526,89948,28911,28,0,"1,8696",0,"0,2174",0,"1,1304",0,...,"0,0435",5,Экзамен,9,1,1,1,4,3,17.06.2022


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 414528 entries, 0 to 414527
Data columns (total 22 columns):
 #   Column                            Non-Null Count   Dtype
---  ------                            --------------   -----
 0   courseid                          414528 non-null  int64
 1   userid                            414528 non-null  int64
 2   num_week                          414528 non-null  int64
 3   s_all                             414528 non-null  int64
 4   s_all_avg                         414528 non-null  str  
 5   s_course_viewed                   414528 non-null  int64
 6   s_course_viewed_avg               414528 non-null  str  
 7   s_q_attempt_viewed                414528 non-null  int64
 8   s_q_attempt_viewed_avg            414528 non-null  str  
 9   s_a_course_module_viewed          414528 non-null  int64
 10  s_a_course_module_viewed_avg      414528 non-null  str  
 11  s_a_submission_status_viewed      414528 non-null  int64
 12  s_a_submission_status_viewe

## Задание 1: Проведите предобработку набора данных

Преобразуйте столбцы, содержащие вещественные значения к типу float.

In [4]:
# ваш код
# Вещественные столбцы записаны с запятой вместо точки (например, "4,5").
# Находим все столбцы со средними значениями и приводим их к типу float.
avg_cols = [c for c in df.columns if c.endswith('_avg')]
for col in avg_cols:
    df[col] = df[col].astype(str).str.replace(',', '.', regex=False).astype('float')

In [5]:
# проверка результата предобработки
print(df[avg_cols].dtypes)
df.head(10)

s_all_avg                           float64
s_course_viewed_avg                 float64
s_q_attempt_viewed_avg              float64
s_a_course_module_viewed_avg        float64
s_a_submission_status_viewed_avg    float64
dtype: object


,courseid,userid,num_week,s_all,s_all_avg,s_course_viewed,s_course_viewed_avg,s_q_attempt_viewed,s_q_attempt_viewed_avg,s_a_course_module_viewed,...,s_a_submission_status_viewed_avg,NameR_Level,Name_vAtt,Depart,Name_OsnO,Name_FormOPril,LevelEd,Num_Sem,Kurs,Date_vAtt
0,71262,34527,6,9,9.0000,4,4.0000,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
1,71262,34527,7,0,4.5000,0,2.0000,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
2,71262,34527,8,0,3.0000,0,1.3333,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
3,71262,34527,9,0,2.2500,0,1.0000,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
4,71262,34527,10,0,1.8000,0,0.8000,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
5,71262,34527,11,1,1.6667,1,0.8333,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
6,71262,34527,12,0,1.4286,0,0.7143,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
7,71262,34527,13,0,1.2500,0,0.6250,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
8,71262,34527,14,0,1.1111,0,0.5556,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022
9,71262,34527,15,1,1.1000,1,0.6000,0,0.0,0,...,0.0,3,Экзамен,22,1,1,1,2,2,18.06.2022


## Примеры визуализаций
Построим столбчатую диаграмму количества студентов по каждому семестру. 
- Метод px.bar() - передаем данные для построения диаграммы
- Функция fig.update_layout() - настройки общего оформления графика (layout): заголовки осей, легенда, фон, отступы, шрифты и другие параметры визуализации.

In [6]:
# Группируем по семестру и считаем количество уникальных студентов
df_grouped = df.groupby("Num_Sem")["userid"].nunique().reset_index()
df_grouped.rename(columns={"userid": "num_students"}, inplace=True)

# Строим столбчатую диаграмму
fig = px.bar(
    df_grouped, 
    x="Num_Sem", 
    y="num_students", 
    text="num_students",  # Подписи с количеством студентов
    color="num_students", 
    title="Количество студентов, изучающих дисциплины в семестр",
    labels={"Num_Sem": "Семестр", "num_students": "Количество студентов"},
    color_continuous_scale="viridis"
)

fig.update_layout(
    xaxis_title="Номер семестра",
    yaxis_title="Количество студентов",
    template="plotly_white"
)

fig.show()

Построим точечную диаграмму количества студентов по каждому семестру. 

px.scatter() - передаем данные для построения диаграммы. 

In [7]:
# Группируем по семестру и считаем количество уникальных студентов
df_grouped = df.groupby("Num_Sem")["userid"].nunique().reset_index()
df_grouped.rename(columns={"userid": "num_students"}, inplace=True)

# Строим точечную диаграмму (scatter plot)
fig = px.scatter(
    df_grouped, 
    x="Num_Sem", 
    y="num_students", 
    text="num_students",  # Подписи значений
    color="num_students", 
    size="num_students",  # Размер точек зависит от количества студентов
    title="Количество студентов, изучающих дисциплины в семестр",
    labels={"Num_Sem": "Семестр", "num_students": "Количество студентов"},
    color_continuous_scale="viridis"
)

fig.update_layout(
    xaxis_title="Номер семестра",
    yaxis_title="Количество студентов",
    template="plotly_white"
)

# Добавляем подписи точек выше самих точек
fig.update_traces(
    textposition='top center',
    marker=dict(
        line=dict(width=1, color='DarkSlateGrey')
    )
)

fig.show()

Построим интерактивный график активности студентов по неделям
Используем plotly.express.area для построения графика изменения активности студентов (s_all) по неделям (num_week). Дополнительно, разделим данные по курсам (Kurs), чтобы увидеть, как активность различается среди студентов разных курсов.
Построить график px.area, где:
- Ось X — num_week (номер недели семестра)
- Ось Y — s_all (общее количество событий)
- Цветовая группировка — Kurs (номер курса студента)
- Добавить анимацию по семестрам (Num_Sem), чтобы увидеть динамику активности студентов разных курсов (свойство -  animation_frame).

In [8]:
# Преобразуем колонки с числами, если они хранятся как строки
numeric_cols = ["s_all", "s_all_avg", "s_course_viewed", "s_course_viewed_avg",
                "s_q_attempt_viewed", "s_q_attempt_viewed_avg", "s_a_course_module_viewed",
                "s_a_course_module_viewed_avg", "s_a_submission_status_viewed", "s_a_submission_status_viewed_avg"]
for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors="coerce")  # Преобразуем к float

# Строим график активности студентов по неделям
fig = px.area(
    df, 
    x="num_week", 
    y="s_all", 
    color="Kurs",  # Разделение по курсам
    line_group="Kurs", 
    title="Динамика активности студентов в течение семестра",
    labels={"num_week": "Неделя", "s_all": "Количество событий", "Kurs": "Курс"},
    animation_frame="Num_Sem"  # Анимация по семестрам
)

fig.show()

## Задание 2. Влияние формы обучения на активность
Решите гипотезу: студенты, обучающиеся на контракте, менее активны на портале по сравнению с бюджетниками.

Что визуализировать: 
Различие в среднем числе взаимодействий между группами ("Name_OsnO") по следующим показателями: "s_all_avg", "s_course_viewed_avg", "s_q_attempt_viewed_avg".
Необходимо использовать в  plotly.express Boxplot распределения активности для бюджетников и контрактников.

In [9]:
# Для сравнения берем итоговое значение по каждой паре «студент – курс»:
# к последней неделе столбцы *_avg содержат среднее число событий в неделю за весь семестр.
metrics = {
    "s_all_avg": "Все события",
    "s_course_viewed_avg": "Просмотры курса",
    "s_q_attempt_viewed_avg": "Просмотры тестов",
}
df_final = df.sort_values("num_week").groupby(["userid", "courseid"]).last().reset_index()

# Предполагаемая кодировка основы обучения: 1 — бюджет, 2 — контракт
df_final["Основа обучения"] = df_final["Name_OsnO"].map({1: "бюджет", 2: "контракт"})

df_box = df_final.melt(
    id_vars=["Основа обучения"], value_vars=list(metrics),
    var_name="metric", value_name="avg_events",
)
df_box["metric"] = df_box["metric"].map(metrics)

fig = px.box(
    df_box, x="Основа обучения", y="avg_events", color="Основа обучения",
    facet_col="metric", points=False,
    category_orders={"Основа обучения": ["бюджет", "контракт"], "metric": list(metrics.values())},
    labels={"avg_events": "Среднее число событий в неделю", "metric": "Показатель"},
    title="Активность на портале: бюджет и контракт",
)
fig.update_yaxes(matches=None, showticklabels=True)  # у каждого показателя своя шкала
fig.update_xaxes(title_text="")
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[-1]))
fig.update_layout(template="plotly_white", showlegend=False)
fig.show()

# Числовое сравнение групп
df_final.groupby("Основа обучения")[list(metrics)].agg(["mean", "median"]).round(2)

s_all_avg        s_course_viewed_avg         \
                     mean median                mean median   
Основа обучения                                               
бюджет              11.68   4.21                2.59   1.25   
контракт            10.04   4.67                1.92   1.00   

                s_q_attempt_viewed_avg         
                                  mean median  
Основа обучения                                
бюджет                            1.72    0.0  
контракт                          1.49    0.0

**Ваш ответ:** гипотеза подтверждается лишь частично.

По средним значениям контрактники действительно менее активны по всем трем показателям: все события – 10,04 против 11,68 в неделю у бюджетников, просмотры курса – 1,92 против 2,59, просмотры тестов – 1,49 против 1,72.

Но распределения сильно скошены вправо: у большинства студентов активность низкая, а небольшая группа очень активных создает длинный «хвост». Поэтому медианы показывают другую картину:
- общее число событий у типичного студента почти одинаково: 4,21 у бюджетников и 4,67 у контрактников (у контрактников даже чуть выше);
- просмотры курса: бюджетники по-прежнему выше (медиана 1,25 против 1,00);
- просмотры тестов: медиана нулевая в обеих группах, то есть больше половины студентов почти не открывают тесты.

Вывод: систематически «менее активными» контрактников назвать нельзя. Различие средних в основном объясняется долей очень активных бюджетников, а типичные студенты двух групп ведут себя похоже.


# Задание 3. Динамика активности студентов по неделям
Гипотеза: Активность студентов увеличивается перед сессией и уменьшается после неё.

Постройте линейный график изменения активности (s_all) по неделям (num_week).
Сравнить активность по курсам (kurs). Все курсы должны отображаться на одной диаграмме. 

In [10]:
# Средняя недельная активность по курсам
df_week = df.groupby(["num_week", "Kurs"], as_index=False)["s_all"].mean()
df_week["Курс"] = df_week["Kurs"].astype(str)
df_total = df.groupby("num_week", as_index=False)["s_all"].mean()

fig = px.line(
    df_week, x="num_week", y="s_all", color="Курс", markers=True,
    labels={"num_week": "Неделя", "s_all": "Среднее число событий на студента"},
    title="Динамика активности студентов по неделям (все курсы)",
)
# среднее по всем курсам
fig.add_trace(go.Scatter(
    x=df_total["num_week"], y=df_total["s_all"], mode="lines",
    name="Все курсы", line=dict(color="black", width=4, dash="dot"),
))
# неделя экзамена по дате Date_vAtt
exam_week = int(pd.to_datetime(df["Date_vAtt"], format="%d.%m.%Y").dt.isocalendar().week.median())
fig.add_vline(x=exam_week, line_dash="dash", line_color="gray",
              annotation_text=f"экзамен (неделя {exam_week})")
fig.update_layout(template="plotly_white")
fig.show()

# Пиковая неделя для каждого курса
df_week.loc[df_week.groupby("Kurs")["s_all"].idxmax(), ["Kurs", "num_week", "s_all"]].round(1)

,Kurs,num_week,s_all
102,2,23,28.6
97,3,22,21.5
116,4,25,19.5
111,5,24,50.5
118,6,25,20.8
113,7,24,9.7


**Ваш ответ:** гипотеза подтверждается.

Средняя недельная активность по всем курсам в недели 6–19 держится около 10 событий на студента, в недели 20–25 (подготовка к сессии) вырастает почти вдвое – до 19,8, а после экзаменов резко падает: 2,5 в недели 26–29 и около 0,5 в недели 27–29. Медианная неделя экзамена по данным Date_vAtt – 25-я, то есть пик активности приходится на последние недели перед ним.

Пиковая неделя по курсам: 2 курс – 23-я, 3 курс – 22-я, 4 курс – 25-я, 5 курс – 24-я (абсолютный максимум, около 50 событий), 6 курс – 25-я, 7 курс – 24-я. У 2–4 курсов в начале семестра (неделя 6) есть небольшой всплеск, после чего активность снижается и держится ровно до начала подготовки к сессии. Курсы 6 и 7 малочисленны, поэтому их линии менее показательны.


## Задание 4: Проанализировать динамику активности студентов с разным уровнем успеваемости (NameR_Level) в течение всего семестра.
Решите гипотезу: успеваемость студентов зависит от динамики их активности на образовательном портале в течение всего семестра. 

Построить линейный график (px.line), где:
X-ось — num_week (номер недели семестра)
Y-ось — s_all (общее количество событий)
Цветовая группировка — NameR_Level (успеваемость: отличники, хорошисты, троечники, двоечники)

In [11]:
level_names = {5: "отличники (5)", 4: "хорошисты (4)", 3: "троечники (3)", 2: "двоечники (2)"}

df_lvl = df.groupby(["num_week", "NameR_Level"], as_index=False)["s_all"].mean()
df_lvl["Успеваемость"] = df_lvl["NameR_Level"].map(level_names)

fig = px.line(
    df_lvl, x="num_week", y="s_all", color="Успеваемость", markers=True,
    category_orders={"Успеваемость": list(level_names.values())},
    labels={"num_week": "Неделя", "s_all": "Среднее число событий на студента"},
    title="Динамика активности студентов с разной успеваемостью",
)
fig.update_layout(template="plotly_white")
fig.show()

# Суммарная активность за семестр на пару «студент – курс»
total = df.groupby(["userid", "courseid"]).agg(total=("s_all", "sum"), grade=("NameR_Level", "first"))
total.groupby("grade")["total"].agg(["mean", "median", "count"]).round(1)

,mean,median,count
grade,,,
2,109.0,25.0,1841
3,179.8,63.0,2939
4,270.3,109.0,5421
5,347.2,154.0,7071


**Ваш ответ:** гипотеза подтверждается: чем выше успеваемость, тем выше активность в течение всего семестра.

В обычные недели (7–19) отличники делают в среднем 11–16 событий в неделю, хорошисты – 7–11, троечники – 4–6, двоечники – 2–5. В среднем за неделю: 14,5 у отличников, 11,3 у хорошистов, 7,5 у троечников и 4,5 у двоечников. За весь семестр на пару «студент – курс» приходится в среднем 347, 270, 180 и 109 событий соответственно (медианы – 154, 109, 63 и 25).

Форма кривых у всех групп похожа: рост перед сессией и спад после нее. Разница в том, как распределяется активность. Отличники и хорошисты работают ровно весь семестр, а троечники и двоечники «включаются» только перед экзаменом: на 25-й неделе троечники (24,5) активнее отличников (15,4). То есть успеваемость связана прежде всего с равномерной работой на портале в течение семестра, а не с авралом перед сессией. Это корреляция, а не доказательство причинной связи.


# Задание 5. Различие в активности бакалавров и магистров
Гипотеза: Магистранты взаимодействуют с курсами иначе, чем бакалавры.

Построить воронку (метод go.Funnel()) активности студентов (s_course_viewed, s_q_attempt_viewed, s_a_submission_status_viewed) в зависимости от уровня образования (leveled).

In [12]:
# Проверяем, какие значения есть в LevelEd
print(df["LevelEd"].unique())  

# закодируем типы обучения
df["LevelEd"] = df["LevelEd"].replace({1: "бакалавриат", 2: "магистратура", 3: "специалитет", 4: "аспирантура"})

# Группируем данные
df_level = df.groupby("LevelEd")[["s_course_viewed", "s_q_attempt_viewed", "s_a_submission_status_viewed"]].mean().reset_index()

print(df_level)

# Строим воронку
fig = go.Figure()
fig.add_trace(go.Funnel(
    name="Бакалавриат",
    y=["Просмотры курса", "Просмотры тестов", "Отправка заданий"],
    x=df_level[df_level["LevelEd"] == "бакалавриат"].iloc[:, 1:].values.flatten()
))
fig.add_trace(go.Funnel(
    name="Магистратура",
    y=["Просмотры курса", "Просмотры тестов", "Отправка заданий"],
    x=df_level[df_level["LevelEd"] == "магистратура"].iloc[:, 1:].values.flatten()
))

fig.update_layout(title="Различие в активности бакалавров и магистров")

# Показываем график
fig.show()

[1 2 3 4]
        LevelEd  s_course_viewed  s_q_attempt_viewed  \
0   аспирантура         1.096264            0.030172   
1   бакалавриат         2.480265            1.719024   
2  магистратура         2.295082            1.638473   
3   специалитет         1.537712            0.598693   

   s_a_submission_status_viewed  
0                      0.909483  
1                      1.617859  
2                      1.385914  
3                      1.008941  


**Ваш ответ:** различия есть, но небольшие.

Среднее недельное число событий на студента у бакалавров и магистров: просмотры курса – 2,48 и 2,30, просмотры тестов – 1,72 и 1,64, отправка заданий – 1,62 и 1,39. Бакалавры чуть активнее на всех трех этапах воронки.

Если смотреть на «конверсию» относительно просмотров курса, то до тестов доходят примерно одинаково (69% у бакалавров и 71% у магистров), а вот отправляют задания реже магистры (60% против 65%). Магистранты чаще ограничиваются просмотром материалов, чем выполнением заданий, но разница невелика, поэтому гипотеза о том, что они взаимодействуют с курсами «иначе», подтверждается лишь слабо.

На воронку выведены только бакалавры и магистры; специалитет и аспирантура в данных тоже есть, но их мало (333 и 58 студентов против 4676 и 1249).


## Задание 6. Различие в активности отличников, хорошистов, троечников и двоечников 
Гипотеза: Студенты с разным уровнем успеваемостью взаимодействуют с курсами иначе.

Построить воронку активности студентов (s_course_viewed, s_q_attempt_viewed, s_a_submission_status_viewed) в зависимости от уровня успеваемости (NameR_Level).

In [13]:
level_names = {5: "отличники (5)", 4: "хорошисты (4)", 3: "троечники (3)", 2: "двоечники (2)"}
stages = ["Просмотры курса", "Просмотры тестов", "Отправка заданий"]
cols = ["s_course_viewed", "s_q_attempt_viewed", "s_a_submission_status_viewed"]

df_perf = df.groupby("NameR_Level")[cols].mean()

fig = go.Figure()
for level in [5, 4, 3, 2]:
    fig.add_trace(go.Funnel(
        name=level_names[level], y=stages, x=df_perf.loc[level].values,
        textinfo="value+percent initial",
    ))
fig.update_layout(title="Различие в активности по уровню успеваемости", template="plotly_white")
fig.show()

df_perf.round(3)

,s_course_viewed,s_q_attempt_viewed,s_a_submission_status_viewed
NameR_Level,,,
2,1.048,0.473,0.622
3,1.627,1.251,1.013
4,2.394,1.670,1.544
5,3.080,2.109,2.008


**Ваш ответ:** гипотеза подтверждается: чем выше успеваемость, тем активнее студенты на каждом этапе воронки.

Среднее недельное число событий (отличники / хорошисты / троечники / двоечники):
- просмотры курса: 3,08 / 2,39 / 1,63 / 1,05;
- просмотры тестов: 2,11 / 1,67 / 1,25 / 0,47;
- отправка заданий: 2,01 / 1,54 / 1,01 / 0,62.

Разрыв растет по мере углубления в курс. Отличники просматривают курс в 2,9 раза чаще двоечников, но тесты открывают в 4,5 раза чаще, а задания отправляют в 3,2 раза чаще. Доля тестов от просмотров курса у отличников – 68%, у двоечников – 45%. Двоечники в основном «заходят и смотрят», но реже переходят к тестам и заданиям.


## Задание 7:  Распределение студентов по кафедрам 
Определим распределение студентов по кафедрам, однаковое ли количество студентов?

Постройте горизонтальную столбчатую диаграмму (bar chart) для сравнения количества студентов на разных кафедрах.
Добавим интерактивность, цветовую палитру и подписи. На диаграмме должны быть выведены названия кафедр (возьмите данную информацию из Task 4.)

In [14]:
# названия кафедр (таблица departments из task 1 / task 4)
departments = pd.read_csv("../task_1_Docker/datasets/departments.csv")

df_dep = df.groupby("Depart")["userid"].nunique().reset_index(name="num_students")
df_dep = df_dep.merge(departments, left_on="Depart", right_on="id").rename(columns={"name": "department"})

fig = px.bar(
    df_dep, x="num_students", y="department", orientation="h",
    text="num_students", color="num_students", color_continuous_scale="viridis",
    labels={"num_students": "Количество студентов", "department": "Кафедра"},
    title="Количество студентов по кафедрам", height=1000,
)
fig.update_yaxes(categoryorder="total ascending", dtick=1)
fig.update_traces(textposition="outside")
fig.update_layout(template="plotly_white", coloraxis_showscale=False)
fig.show()

df_dep.sort_values("num_students", ascending=False)["num_students"].describe().round(1)

count     43.0
mean     146.9
std       89.4
min        5.0
25%       83.5
50%      125.0
75%      220.0
max      362.0
Name: num_students, dtype: float64

**Ваш ответ:** нет, количество студентов по кафедрам распределено неравномерно.

Число студентов колеблется от 5 (УСиБА) до 362 (ДиСО) при медиане 125 и среднем около 147. Больше всего студентов на кафедрах ДиСО (362), РМПИ (327), МиХТ (313), ГМДиОПИ (293) и ЛиП (248). Меньше всего – на УСиБА (5), ИТМ (7), Физкульт. (11), ЦДОМ (11) и Физики (50).

Студент учитывается на каждой кафедре, чьи курсы он изучает, поэтому сумма по кафедрам больше общего числа студентов (6316).


## Задание 8: Зависимость среднего балла от количества студентов на кафедре
Гипотеза: количество студентов, закрепленных на отдельной кафедре влияет на средний балл успеваемости по кафедрам.

Построить диаграмму рассевания по количеству студентов, среднему баллу, наванию кафедры. Параметр  size='num_students' - количество студентов. 

In [15]:
# Оценка выставляется один раз за курс, поэтому убираем недельные дубли строк
df_grade = df.drop_duplicates(["userid", "courseid"])
df_dept = df_grade.groupby("Depart").agg(
    num_students=("userid", "nunique"), avg_grade=("NameR_Level", "mean")
).reset_index()
df_dept = df_dept.merge(departments, left_on="Depart", right_on="id").rename(columns={"name": "department"})

fig = px.scatter(
    df_dept, x="num_students", y="avg_grade", size="num_students",
    color="avg_grade", color_continuous_scale="viridis",
    hover_name="department", text="department",
    labels={"num_students": "Количество студентов", "avg_grade": "Средний балл"},
    title="Зависимость среднего балла от количества студентов на кафедре",
)
# линия тренда (МНК)
k, b = np.polyfit(df_dept["num_students"], df_dept["avg_grade"], 1)
xs = np.array([df_dept["num_students"].min(), df_dept["num_students"].max()])
fig.add_trace(go.Scatter(x=xs, y=k * xs + b, mode="lines", name="тренд",
                         line=dict(color="gray", dash="dash")))
fig.update_traces(textposition="top center", selector=dict(mode="markers+text"))
fig.update_layout(template="plotly_white", coloraxis_showscale=False)
fig.show()

print("Корреляция Пирсона :", round(df_dept["num_students"].corr(df_dept["avg_grade"]), 3))
print("Корреляция Спирмена:", round(df_dept["num_students"].corr(df_dept["avg_grade"], method="spearman"), 3))

Корреляция Пирсона : -0.345
Корреляция Спирмена: -0.327


**Ваш ответ:** гипотеза подтверждается слабо: зависимость есть, но она умеренная.

Корреляция между числом студентов и средним баллом по кафедрам отрицательная: коэффициент Пирсона –0,35, Спирмена –0,33. Чем больше студентов на кафедре, тем в среднем ниже балл: у кафедр с числом студентов выше медианы средний балл 3,95, у остальных – 4,22. Самые низкие средние баллы у УиИС (3,47, 80 студентов), АЭПиМ (3,55, 241) и ТОМ (3,57, 226), а самые высокие (до 5,0) – у очень малых кафедр: УСиБА (5 студентов) и Физкульт. (11).

Но r² ≈ 0,12, то есть размер кафедры объясняет лишь около 12% различий в среднем балле. На результат сильно влияют специфика дисциплин и случайный разброс на малых кафедрах, поэтому делать вывод о причинной связи нельзя.
